src/_pytest/nodes.py

```diff
@@ -135,18 +137,48 @@ def __repr__(self):
         return "<%s %r>" % (self.__class__.__name__, getattr(self, "name", None))
 
     def warn(self, code, message):
-        """ generate a warning with the given code and message for this
-        item. """
+        """
+        .. deprecated:: 3.8
+
+            Use :meth:`Node.std_warn <_pytest.nodes.Node.std_warn>` instead.
+
+        Generate a warning with the given code and message for this item.
+        """
+        from _pytest.warning_types import RemovedInPytest4Warning
+
+        self.std_warn(
+            "Node.warn has been deprecated, use Node.std_warn instead",
+            RemovedInPytest4Warning,
+        )
+
         assert isinstance(code, str)
-        fslocation = getattr(self, "location", None)
-        if fslocation is None:
-            fslocation = getattr(self, "fspath", None)
+        fslocation = get_fslocation_from_item(self)
         self.ihook.pytest_logwarning.call_historic(
             kwargs=dict(
                 code=code, message=message, nodeid=self.nodeid, fslocation=fslocation
             )
         )
 
+    def std_warn(self, message, category=None):
```

Comment:
we should do better than the stdlib here any only support warning instances - the duality generates dissonance